### PR TITLE
Mathv3

### PR DESCRIPTION
**Pull request by dgrat**
@lucasdemarchi @guludo 
This is the new version of the AP_Math module. I think, I fixed all float to double promotions on my site thanks to @lucasdemarchi. All (few) remaining float to double conversions are because the authors decided to mix floats with ints in the AP code. Theoretically, I could fix them by hand if someone wishes, so AP would be entirely free of doubles, but I think they are not related to this module. 
I decided to split the implementation for PX4 and other platforms for some functions to guarantee the return of a float (not double) and to avoid an over-complicated implementation for the other platforms (Linux).  
I followed @guludo 's advice to replace the return type of the converter functions with a Template argument. I left the zero check, because it would anyway crash if there is a zero division for integers. 

Best, Daniel

### CODE REVIEW

**Comment by dgrat** [comment]
I splitted the platform dependent implementation now. 
The PX4 library will not generate anymore double promotion warnings, because I made promotions impossible. The cause is/was the excessive use of integers in AP together with functions expecting floats. In case of templates the compiler inserts doubles, because the integer part of a float is smaller than 32 bit.

**Comment by dgrat** [comment]
* I re-defined the M_*_PI_* macros for PX4 as float. 
  * After this, all the double promotion possibilities should be removed

**Comment by lucasdemarchi** [comment]
@dgrat thanks for the work on this, it's taking shape to be mergeable. Sorry for not reviewing this, I've just got too much on my plate lately. Would be nice to have other people to review as well. I'll try to review it next week. Thanks.

File: APMrover2/test.cpp [hunk 1]
```diff
@@ -402,7 +402,7 @@ int8_t Rover::test_mag(uint8_t argc, const Menu::arg *argv)
             if (compass.healthy()) {
                 const Vector3f mag_ofs = compass.get_offsets();
                 const Vector3f mag = compass.get_field();
-                cliSerial->printf("Heading: %d, XYZ: %.0f, %.0f, %.0f,\tXYZoff: %6.2f, %6.2f, %6.2f\n",
+                cliSerial->printf("Heading: %f, XYZ: %.0f, %.0f, %.0f,\tXYZoff: %6.2f, %6.2f, %6.2f\n",
```
Comment (magicrub) [inline APMrover2/test.cpp]: Why is this %f now?

File: ArduCopter/control_autotune.cpp [hunk 2]
```diff
@@ -301,7 +301,7 @@ void Copter::autotune_run()
         pos_control.relax_alt_hold_controllers(get_throttle_pre_takeoff(channel_throttle->control_in)-throttle_average);
     }else{
         // check if pilot is overriding the controls
-        if (!is_zero(target_roll) || !is_zero(target_pitch) || !is_zero(target_yaw_rate) || !is_zero(target_climb_rate)) {
+        if (!is_zero(target_roll) || !is_zero(target_pitch) || !is_zero(target_yaw_rate) || target_climb_rate != 0) {
```
Comment (magicrub) [inline ArduCopter/control_autotune.cpp]: Doesn't this create a warning?

File: ArduCopter/control_guided.cpp [hunk 3]
```diff
@@ -215,7 +215,7 @@ void Copter::guided_set_angle(const Quaternion &q, float climb_rate_cms)
     q.to_euler(guided_angle_state.roll_cd, guided_angle_state.pitch_cd, guided_angle_state.yaw_cd);
     guided_angle_state.roll_cd = ToDeg(guided_angle_state.roll_cd) * 100.0f;
     guided_angle_state.pitch_cd = ToDeg(guided_angle_state.pitch_cd) * 100.0f;
-    guided_angle_state.yaw_cd = wrap_180_cd_float(ToDeg(guided_angle_state.yaw_cd) * 100.0f);
+    guided_angle_state.yaw_cd = wrap_180_cd(ToDeg(guided_angle_state.yaw_cd) * 100.0f);
```
Comment (magicrub) [inline ArduCopter/control_guided.cpp]: I'm seeing a lot of integer math using *100.0f which seems unnecessary

File: ArduCopter/control_rtl.cpp [hunk 4]
```diff
@@ -85,7 +85,7 @@ void Copter::rtl_climb_start()
     wp_nav.wp_and_spline_init();
 
     // RTL_SPEED == 0 means use WPNAV_SPEED
-    if (!is_zero(g.rtl_speed_cms)) {
+    if (!is_zero((float)g.rtl_speed_cms)) {
```
Comment (magicrub) [inline ArduCopter/control_rtl.cpp]: Unnecessary cast, just don't use iszero.

**Comment by magicrub** [comment]
I'm seeing a lot of changes needed, all small stuff but they are legitimate bugs. Daniel, have you run this in the simulator yet?

File: APMrover2/test.cpp (hunk 1 above)
Comment (dgrat) [inline APMrover2/test.cpp]: Because heading is a float :p

File: ArduCopter/control_autotune.cpp (hunk 2 above)
Comment (dgrat) [inline ArduCopter/control_autotune.cpp]: No, it was an int :P

File: ArduCopter/control_guided.cpp (hunk 3 above)
Comment (dgrat) [inline ArduCopter/control_guided.cpp]: yes

File: ArduCopter/control_rtl.cpp (hunk 4 above)
Comment (dgrat) [inline ArduCopter/control_rtl.cpp]: No, I think it was AP_* type

**Comment by dgrat** [comment]
> While instead of if
No

**Comment by dgrat** [comment]
Hi there,
I will start testing the commits. Guess, it will take a while.
@magicrub I will try to include some of the recommendations. 
I also think that some of them need further testing, but some of the changes are included by some reasons. 

Best, Daniel

**Comment by lucasdemarchi** [comment]
@dgrat any chance you could leverage our integration with gtest for unit tests?

File: ArduCopter/control_rtl.cpp (hunk 4 above)
Comment (lucasdemarchi) [inline ArduCopter/control_rtl.cpp]: It's an AP_Int16. This should be changed to

```c++
if (g.rtl_speed_cms.get() != 0)
```

Depending on the case (not here)... calling the cast_to_float() method is another option, too.

**Comment by dgrat** [comment]
Yes, i think so

File: ArduCopter/control_rtl.cpp (hunk 4 above)
Comment (dgrat) [inline ArduCopter/control_rtl.cpp]: I will change it, but i want to add that there have been more of such things in the code.

**Comment by dgrat** [comment]
@lucasdemarchi 
I guess, I have to read something about gtest first.
Not sure atm how to include sane unit tests with WAF.

**Comment by dgrat** [comment]
root@debian:/home/dgrat/ardupilot# ./waf configure --board 'linux' build
bash: ./waf: Datei oder Verzeichnis nicht gefunden                                                                                                                                                                                                                             
root@debian:/home/dgrat/ardupilot# python ../waf configure --board 'linux' build                                                                                                                                                                                               
Setting top to                           : /home/dgrat/ardupilot                                                                                                                                                                                                               
Setting out to                           : /home/dgrat/ardupilot/build                                                                                                                                                                                                         
Setting board to                         : linux                                                                                                                                                                                                                               
Checking for 'g++' (C++ compiler)        : g++                                                                                                                                                                                                                                 
Checking for 'gcc' (C compiler)          : gcc                                                                                                                                                                                                                                 
Could not load the Waf tool 'clang_compilation_database' from ['Tools/ardupilotwaf/', '/home/dgrat/.waf-1.8.17-daa67386b71d29ad7cda321d44eb497a', '/home/dgrat', '/usr/lib/python2.7', '/usr/lib/python2.7/plat-x86_64-linux-gnu', '/usr/lib/python2.7/lib-tk', '/usr/lib/python2.7/lib-old', '/usr/lib/python2.7/lib-dynload', '/usr/local/lib/python2.7/dist-packages', '/usr/lib/python2.7/dist-packages', '/usr/lib/python2.7/dist-packages/PILcompat', '/usr/lib/python2.7/dist-packages/gtk-2.0', '/usr/lib/pymodules/python2.7']                       
No module named clang_compilation_database                                                                                                                                                                                                                                     
(complete log in /home/dgrat/ardupilot/build/config.log)

**Comment by dgrat** [comment]
@lucasdemarchi 
I wrote some field tests, which might be not entirely correct. Additionally, I found a mistake in my and the old wrap functions. E.g. wrap_w180(180) does not return a 180 but a -180. The same for wrap_PI(). This is not correct, but probably doesn't matter. 
Additionally, I have the feeling that arming doesn't work because of the new wrap_functions?! I am not sure whether its really the commit or something else, but I think the wrap_functions must be entirely correct now. In contrast, the old wrap function converge earlier to wrong results. Can someone, maybe have a look at this as well?!

**Comment by lucasdemarchi** [comment]
> @lucasdemarchi 
> I guess, I have to read something about gtest first.
> Not sure atm how to include sane unit tests with WAF.

Have you checked https://github.com/diydrones/ardupilot/blob/master/README-WAF.md ?

Basically you have to use waf we have in our repository:

```
$ ./modules/waf/waf-light configure --board linux
Setting top to                           : /home/lucas/p/dronecode/ardupilot 
Setting out to                           : /home/lucas/p/dronecode/ardupilot/build 
Setting board to                         : linux 
Checking for 'g++' (C++ compiler)        : g++ 
Checking for 'gcc' (C compiler)          : gcc 
Checking for program 'python'            : /usr/bin/python2 
Checking for python version              : (2, 7, 10, 'final', 0) 
Checking for program 'git'               : /usr/bin/git 
Checking for program 'cmake'             : /usr/bin/cmake 
Checking for program 'ninja, ninja-build' : /usr/bin/ninja-build 
Benchmarks                                : enabled 
Unit tests                                : enabled 
'configure' finished successfully (0.256s)
```

The readme file shows how to add an alias so you don't have to type the `modules/waf/...` thing.

**Comment by dgrat** [comment]
@lucasdemarchi 
WAF seems to work now :D

**Comment by dgrat** [comment]
root@debian:/home/dgrat/ardupilot# ./modules/waf/waf-light check-all
    Waf: Entering directory `/home/dgrat/ardupilot/build/sitl'
    [1515/1564] Creating Google Benchmark
    [1517/1564] Linking build/sitl/examples/eulers
    Command '/usr/bin/cmake --build /home/dgrat/ardupilot/build/sitl/gbenchmark_build --target install' returned 1

    die: error: unable to read configuration file
    Waf: Leaving directory `/home/dgrat/ardupilot/build/sitl'
    Build failed
     -> task in 'benchmarks/benchmark_matrix' failed (exit status 1): 
            {task 140340818648656: gbenchmark_build  -> benchmark.h,macros.h,benchmark_api.h,reporter.h,libbenchmark.a}
    ''

**Comment by dgrat** [comment]
@lucasdemarchi 
Initial unit tests should work now :D

**Pull request closed without merging**